### PR TITLE
[Hardware][CPU] Update torch 2.5

### PR DESCRIPTION
**Pull request by bigPYJ1151**
- Update torch 2.5
- Enable FP16 data type

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by zhouyuan** [comment]
related
https://github.com/vllm-project/vllm/pull/9911

File: Dockerfile.cpu [hunk 1]
```diff
@@ -22,7 +22,8 @@ ENV LD_PRELOAD="/usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.so.4:/usr/local/li
 
 RUN echo 'ulimit -c 0' >> ~/.bashrc
 
-RUN pip install intel_extension_for_pytorch==2.4.0
+# RUN pip install intel_extension_for_pytorch==2.4.0
```
Comment (jikunshang) [inline Dockerfile.cpu:25]: please remove this

**Review by jikunshang — APPROVED** [review]

**Comment by Isotr0py** [comment]
Perhaps #10021 will have conflicts with the Ray installation checking in this PR?

**Comment by bigPYJ1151** [comment]
> Perhaps #10021 will have conflicts with the Ray installation checking in this PR?

Thanks for the hint, revert it.

File: vllm/executor/cpu_executor.py [hunk 2]
```diff
@@ -316,11 +314,6 @@ async def check_health_async(self) -> None:
 
 
 def _verify_and_get_model_config(config: ModelConfig) -> ModelConfig:
-    if config.dtype == torch.float16:
-        logger.warning("float16 is not supported on CPU, casting to bfloat16.")
-        config.dtype = torch.bfloat16
-    # Reminder: Please update docs/source/serving/compatibility_matrix.rst
-    # If the feature combo become valid
```
Comment (DarkLight1337) [inline vllm/executor/cpu_executor.py:323]: We should keep these comments.

File: .buildkite/run-cpu-test.sh [hunk 3]
```diff
@@ -46,12 +46,12 @@ docker exec cpu-test bash -c "
...
+  python3 -m vllm.entrypoints.openai.api_server --model facebook/opt-125m --dtype bfloat16 & 
   timeout 600 bash -c 'until curl localhost:8000/v1/models; do sleep 1; done' || exit 1
   python3 benchmarks/benchmark_serving.py \
     --backend vllm \
     --dataset-name random \
     --model facebook/opt-125m \
     --num-prompts 20 \
     --endpoint /v1/completions \
-    --tokenizer facebook/opt-125m"
+    --tokenizer facebook/opt-125m"
```
Comment (DarkLight1337) [inline .buildkite/run-cpu-test.sh:57]: We should keep this newline.

File: csrc/cpu/dnnl_helper.hpp [hunk 4]
```diff
@@ -32,6 +33,11 @@ struct DNNLType<c10::BFloat16> {
   static constexpr dnnl::memory::data_type type = dnnl::memory::data_type::bf16;
 };
 
+template <>
+struct DNNLType<c10::Half> {
+  static constexpr dnnl::memory::data_type type = dnnl::memory::data_type::f16;
+};
+
```
Comment (Isotr0py) [inline csrc/cpu/dnnl_helper.hpp:40]: Seems that DNNL supports f16 as well, could we update the IPEX quantization support dtype to use AWQ quantization with fp16?
https://github.com/vllm-project/vllm/blob/ad23318928d40ef7ac969451afa0dc198428c04b/vllm/model_executor/layers/quantization/ipex_quant.py#L55-L58

**Review by Isotr0py — APPROVED** [review]
Overall looks reasonable to me. Just have a minor question :)

File: csrc/cpu/dnnl_helper.hpp (hunk 4 above)
Comment (bigPYJ1151) [reply csrc/cpu/dnnl_helper.hpp:40]: Updated.

**Review by mgoin — COMMENTED** [review]
Why is the requirement for __AVX512FP16__ removed? What was it previously used for if we didn't have a dispatch case for half?

File: .buildkite/run-cpu-test.sh [hunk 5]
```diff
@@ -46,7 +46,7 @@ docker exec cpu-test bash -c "
 docker exec cpu-test bash -c "
   export VLLM_CPU_KVCACHE_SPACE=10 
   export VLLM_CPU_OMP_THREADS_BIND=48-92 
-  python3 -m vllm.entrypoints.openai.api_server --model facebook/opt-125m & 
+  python3 -m vllm.entrypoints.openai.api_server --model facebook/opt-125m --dtype bfloat16 &
```
Comment (WoosukKwon) [inline .buildkite/run-cpu-test.sh:49]: Why do we need bfloat16 here? It looks contradictory to this PR.

**Comment by bigPYJ1151** [comment]
@mgoin Previously we have some FP16 vector type using **AVX512FP16**. Because only the cast intrinsic between FP16 and FP32 is required, so the **F16C** is enough and available on most of x86 platforms.

File: .buildkite/run-cpu-test.sh (hunk 5 above)
Comment (bigPYJ1151) [reply .buildkite/run-cpu-test.sh]: No special reason, changed to ```half```.

**Comment by bigPYJ1151** [comment]
Hi @Isotr0py @DarkLight1337 @mgoin @WoosukKwon , maybe we can mark this PR ready if there is no further comments, thanks :)

**Review by DarkLight1337 — APPROVED** [review]
Sure, let's get this merged.

**Review by mgoin — APPROVED** [review]

**Pull request merged**